src/System.Net.Sockets/src/System/Net/Sockets/IOControlKeepAlive.cs

```diff
@@ -0,0 +1,103 @@
+
+        public static bool TryGet(SafeSocketHandle handle, SocketOptionLevel optionLevel, SocketOptionName optionName, out int optionValue)
+        {
+            if (IsNotNeeded(optionLevel, optionName))
+            {
+                optionValue = 0;
+                return false;
+            }
+                
+            IOControlKeepAlive ioControlKeepAlive;
+            s_socketKeepAliveCwt.TryGetValue(handle, out ioControlKeepAlive);
+            optionValue = optionName == SocketOptionName.TcpKeepAliveTime ?
+                MillisecondsToSeconds(ioControlKeepAlive?._timeMs ?? WindowsDefaultTimeMs) :
+                MillisecondsToSeconds(ioControlKeepAlive?._intervalMs ?? WindowsDefaultIntervalMs);
+            return true;
+        }
+
+        public static SocketError? Set(SafeSocketHandle handle, SocketOptionLevel optionLevel, SocketOptionName optionName, int optionValue)
+        {
+            if (IsNotNeeded(optionLevel, optionName))
+                return null;
+
+            IOControlKeepAlive ioControlKeepAlive = s_socketKeepAliveCwt.GetOrCreateValue(handle);
+            if (optionName == SocketOptionName.TcpKeepAliveTime)
+                ioControlKeepAlive._timeMs = SecondsToMilliseconds(optionValue);
+            else
+                ioControlKeepAlive._intervalMs = SecondsToMilliseconds(optionValue);
+
+            int realOptionLength = 0;
+            return SocketPal.WindowsIoctl(handle, unchecked((int)IOControlCode.KeepAliveValues), ioControlKeepAlive.ToByteArray(), null, out realOptionLength);
+        }
+
+        private static bool SupportsKeepAliveViaSocketOption()
+        {
+            using (Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                int time = MillisecondsToSeconds(WindowsDefaultTimeMs);
+                SocketError timeErrCode = Interop.Winsock.setsockopt(
+                    socket.SafeHandle,
+                    SocketOptionLevel.Tcp,
+                    SocketOptionName.TcpKeepAliveTime,
+                    ref time,
+                    sizeof(int));
+                int interval = MillisecondsToSeconds(WindowsDefaultIntervalMs);
+                SocketError intervalErrCode = Interop.Winsock.setsockopt(
+                    socket.SafeHandle,
+                    SocketOptionLevel.Tcp,
+                    SocketOptionName.TcpKeepAliveInterval,
+                    ref time,
+                    sizeof(int));
+                return timeErrCode == SocketError.Success && intervalErrCode == SocketError.Success;
+            }
+        }
+
+        private static bool IsNotNeeded(SocketOptionLevel optionLevel, SocketOptionName optionName)
+        {
+            return s_supportsKeepAliveViaSocketOption ||
+                optionLevel != SocketOptionLevel.Tcp ||
+                (optionName != SocketOptionName.TcpKeepAliveTime && optionName != SocketOptionName.TcpKeepAliveInterval);
```

Comment:
The option name/level checks should be moved it out of this type into the Socket pal call sites. This type should only ever be touched when actually setting these options.